File: nanopc/src/Car.cpp
#include "Car.h"

Car::Car()
    : serial_()
    , packet_()
    , tranSpeed_(0)
    , rotaSpeed_(0)
{
}

Car::~Car()
{
    move(0, 0);
}

void Car::init()
{
    serial_.openPort("/dev/ttySAC2");
    serial_.setup(115200, 0, 8, 1, 'N');
}

void Car::move(Speed tranSpeed, Speed rotaSpeed)
{
    tranSpeed_ = tranSpeed;
    rotaSpeed_ = rotaSpeed;
    updatePacket();
}

void Car::updatePacket()
{
    setPacketHead();
    setPacketData();
    setPacketCheck();
}

void Car::setPacketHead()
{
    packet_[0] = 0xAA;
}

void Car::setPacketData()
{
    packet_[1] = tranSpeed_ & 0xFF;
    packet_[2] = tranSpeed_ >> 8;
    packet_[3] = rotaSpeed_ & 0xFF;
    packet_[4] = rotaSpeed_ >> 8;
}

void Car::setPacketCheck()
{
    uint8_t sum = 0;
    for(int i = 0; i < kPacketSize_; ++i) {
        sum += packet_[i];
    }
    packet_[kPacketSize_ - 1] = sum;
}

File: nanopc-py/Car.py
import serial

class Car :
    def __init__(self, dev, baudRate) :
        self.__serial = serial.Serial(dev, baudRate)
        self.__kPacketSize = 6
        self.__packet = bytearray(self.__kPacketSize)
        self.__tranSpeed = 0
        self.__rotaSpeed = 0

    def __del__(self) :
        self.stop()
        self.__serial.close()

    def move(self, tranSpeed, rotaSpeed) :
        self.__tranSpeed = tranSpeed
        self.__rotaSpeed = rotaSpeed
        self.__updatePacket()
        self.__sendPacket()

    def stop(self) :
        self.move(0, 0)

    def __setPacketHead(self) :
        self.__packet[0] = 0xAA

    def __setPacketData(self) :
        self.__packet[1] = self.__tranSpeed & 0xFF
        self.__packet[2] = self.__tranSpeed >> 8 & 0xFF
        self.__packet[3] = self.__rotaSpeed & 0xFF
        self.__packet[4] = self.__rotaSpeed >> 8 & 0xFF

    def __setPacketCheck(self) :
        sum = 0
        for i in range(self.__kPacketSize-1) :
            sum += self.__packet[i]
        self.__packet[self.__kPacketSize-1] = sum & 0xFF
    
    def __updatePacket(self) :
        self.__setPacketHead()
        self.__setPacketData()
        self.__setPacketCheck()

    def __sendPacket(self) :
        self.__serial.write(self.__packet)
